main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -430,6 +437,90 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
     }
   }
 
+  def upgradeEffects(flix: Flix): Result[Unit, BootstrapError] = {
+    if (!isProjectMode) {
+      return Err(BootstrapError.FileError("No 'flix.toml' found. Refusing to run 'eff-upgrade'"))
+    }
+    Steps.updateStaleSources(flix)
+    for {
+      root <- Steps.check(flix)
+    } yield {
+      try {
+        if (!Files.exists(Bootstrap.getEffectLockFile(projectPath))) {
+          return Err(BootstrapError.FileError("No 'effects.lock' file found. Refusing to run 'eff-upgrade'."))
+        }
+      } catch {
+        case e: Exception => return Err(BootstrapError.FileError(s"IO error: ${e.getMessage}"))
+      }
+      val json =
+        try {
+          Files.readString(Bootstrap.getEffectLockFile(projectPath))
+        } catch {
+          case _: Exception => return Err(BootstrapError.FileError("Unable to read 'effects.lock'. Refusing to run 'eff-upgrade'."))
+        }
+
+      EffectLock.deserialize(json) match {
+        case Err(e) => return Err(BootstrapError.FileError(e))
+        case Ok((defs, sigs)) =>
+          // 3. Check effect lock
+          // N.B.: Map to strings to remove different internal sym ids
+          val strToLockedDefs = defs.map {
+            case (sym, scheme) => sym.toString -> scheme
+          }
+          val strToLockedSigs = sigs.map {
+            case (sym, scheme) => sym.toString -> scheme
+          }
+          val strToNewDefs = root.defs.map {
+            case (sym, scheme) => sym.toString -> scheme
+          }
+          val strToNewSigs = root.sigs.map {
+            case (sym, scheme) => sym.toString -> scheme
+          }
+          val defSchemes = strToLockedDefs.collect {
+            case (sym, originalScheme) if strToNewDefs.contains(sym) =>
+              val upgradeScheme = strToNewDefs(sym)
+              sym -> (originalScheme, upgradeScheme.spec.declaredScheme)
+          }
+          val sigSchemes = strToLockedSigs.collect {
+            case (sym, originalScheme) if strToNewSigs.contains(sym) =>
+              val upgradeScheme = strToNewSigs(sym)
+              sym -> (originalScheme, upgradeScheme.spec.declaredScheme)
+          }
+          val useGraph = ListMap.from(UseGraph.computeGraph(root).map {
+            case (src, UseGraph.UsedSym.DefnSym(dst)) => src.toString -> dst.loc
+            case (src, UseGraph.UsedSym.SigSym(dst)) => src.toString -> dst.loc
+          })
+          val defUpgradeErrors = defSchemes.map {
+            case (sym, (original, upgrade)) =>
+              val uses = useGraph.get(sym)
+              if (uses.isEmpty || EffectUpgrade.isEffSafeUpgrade(original, upgrade)(flix)) {
```

Comment:
this long sequence certainly feels like it could be refactored into something prettier. Lots of repetition and we are only really interested in the output, not the intermediate steps.